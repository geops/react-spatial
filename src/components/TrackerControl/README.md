
The following example demonstrates the use of TrackerControl.

```jsx
import React from 'react';
import { Layer, TrajservLayer } from 'mobility-toolbox-js/ol';
import Tile from 'ol/layer/Tile';
import OSM from 'ol/source/OSM';
import BasicMap from 'react-spatial/components/BasicMap';
import TrackerControl from 'react-spatial/components/TrackerControl';

// The `apiKey` used here is for demonstration purposes only.
// Please get your own api key at https://developer.geops.io/.
const trackerLayer = new TrajservLayer({
  apiKey: window.apiKey,
  live: false
});

const layers = [
  new Layer({
    olLayer: new Tile({
      source: new OSM(),
    }),
  }),
  trackerLayer,
];

<>
  <BasicMap center={[951560, 6002550]} zoom={14} layers={layers} tabIndex={0} />
  <TrackerControl trackerLayer={trackerLayer} />
</>
```
