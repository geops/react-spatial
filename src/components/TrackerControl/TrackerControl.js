import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import { IoIosSpeedometer } from 'react-icons/io';
import { FaPlay, FaForward, FaBackward, FaRegDotCircle } from 'react-icons/fa';

import TrackerLayer from '../../layers/TrackerLayer';

const increaseSpeed = (speed) => {
  let delta = 0.1;
  if (speed >= 1) {
    delta = 1;
  }
  if (speed >= 10) {
    delta = 5;
  }
  const nextSpeed = speed + delta;
  return nextSpeed > 30 ? speed : nextSpeed;
};

const decreaseSpeed = (speed) => {
  let delta = 0.1;
  if (speed > 1) {
    delta = 1;
  }
  if (speed > 10) {
    delta = 5;
  }
  const nextSpeed = speed - delta;
  if (nextSpeed < 0.1) {
    return speed;
  }
  return nextSpeed;
};

const defaultRenderButton = (icon, onClick, title) => {
  return (
    <div
      aria-label={title}
      role="button"
      onClick={onClick}
      onKeyPress={onClick}
      className="rt-control-button"
      tabIndex={0}
      title={title}
    >
      {icon}
    </div>
  );
};

/**
 * TrackerControl allows the user to control the speed of a trackerLayer.
 */
function TrackerControl({
  className,
  iconDateReset,
  iconSpeedDown,
  iconSpeedReset,
  iconSpeedUp,
  iconSpeed,
  renderButton,
  trackerLayer,
}) {
  const [speed, setSpeed] = useState(1);

  const onSpeedChange = (newSpeed) => {
    trackerLayer.setSpeed(newSpeed);
  };

  const resetDate = () => {
    trackerLayer.setCurrTime(new Date());
  };

  useEffect(() => {
    onSpeedChange(speed);
  }, [speed, onSpeedChange]);

  return (
    <div className={className}>
      {renderButton(iconDateReset, () => resetDate(), 'reset date')}
      {renderButton(
        iconSpeedDown,
        () => setSpeed(decreaseSpeed(speed)),
        'speed down',
      )}
      {renderButton(iconSpeedReset, () => setSpeed(1), 'speed reset')}
      {renderButton(
        iconSpeedUp,
        () => setSpeed(increaseSpeed(speed)),
        'speed up',
      )}
      <div className="rt-tracker-speed">
        {iconSpeed}
        {`${speed < 1 ? speed.toFixed(1) : speed}`}
      </div>
    </div>
  );
}

TrackerControl.propTypes = {
  /**
   * CSS class of the tracker control.
   */
  className: PropTypes.string,

  /**
   * Icon of the date reset button.
   */
  iconDateReset: PropTypes.element,

  /**
   * Icon of the speed down button.
   */
  iconSpeedDown: PropTypes.element,

  /**
   * Icon of the speed up button.
   */
  iconSpeedUp: PropTypes.element,

  /**
   * Icon of the speed reset button.
   */
  iconSpeedReset: PropTypes.element,

  /**
   * Icon speed.
   */
  iconSpeed: PropTypes.element,

  /**
   * Render function for buttons.
   */
  renderButton: PropTypes.func,

  /**
   * Trackerlayer.
   */
  trackerLayer: PropTypes.instanceOf(TrackerLayer).isRequired,
};

TrackerControl.defaultProps = {
  className: 'rt-tracker-control',
  iconDateReset: <FaRegDotCircle />,
  iconSpeedDown: <FaBackward />,
  iconSpeedReset: <FaPlay />,
  iconSpeedUp: <FaForward />,
  iconSpeed: <IoIosSpeedometer />,
  renderButton: defaultRenderButton,
};

export default TrackerControl;
