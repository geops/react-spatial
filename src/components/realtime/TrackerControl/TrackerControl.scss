.rt-tracker-control {
  display: flex;
  align-items: center;

  .rt-control-button {
    border: 1px solid #ccc;
    color: #333;
    background-color: #fff;
    padding: 8px 12px 4px 12px;
    margin: 2px;
    border-radius: 4px;

    &:hover {
      cursor: pointer;
      background-color: #f5f5f5;
    }
  }

  .rt-tracker-speed {
    display: flex;
    align-items: center;
    padding-left: 15px;

    svg {
      height: 1.2em;
      width: 1.2em;
      padding-right: 5px;
    }
  }
}
