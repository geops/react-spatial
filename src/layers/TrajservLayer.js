import qs from 'query-string';
import Feature from 'ol/Feature';
import { transform as transformCoords } from 'ol/proj';
import { buffer, getWidth } from 'ol/extent';
import { Point, MultiPoint, LineString } from 'ol/geom';
import { Style, Fill, Stroke, Circle } from 'ol/style';
import TrackerLayer from './TrackerLayer';
import { getDateString, getUTCTimeString } from '../utils/TimeUtils';
import {
  getRadius,
  getBgColor,
  getDelayColor,
  getDelayText,
  getTextColor,
  getTextSize,
} from '../config/tracker';

/**
 * Responsible for loading tracker data from Trajserv.
 * @class
 * @inheritDoc
 * @param {Object} [options] Layer options.
 * @param {string} [url = https://api.geops.io/tracker] Tracker url.
 * @param {string} options.apiKey Access key for [geOps services](https://developer.geops.io/).
 * @param {number} options.delayDisplay delay from which the time is always display on the feature (in milliseconds).
 * @param {Array.<string>|string} options.regexPublishedLineName Regex filter for line name. This filter has a higher prio over publishedLineName.
 * @param {Array.<string>|string} options.publishedLineName Filter by line name, string: 'ICE',  list: 's1,s2,s9,s10,s15'
 * @param {Array.<string>|string} options.tripNumber Filter by trip number, bus in zurich: '2068', list of buses in Zurich: '2068,3003,3451,3953'
 * @param {Array.<string>|string} options.operator Filter by operator, string: 'sbb', list: '(vbz\|zsg)'
 */
class TrajservLayer extends TrackerLayer {
  /**
   *  Translate the response date object into a readable object.
   * @returns {array<Date>}
   * @private
   */
  static translateDates(dates) {
    const newDates = [];

    for (let i = 0; i < dates.length; i += 1) {
      const { d: day, m: month, y: year } = dates[i];
      newDates.push({
        day,
        month,
        year,
      });
    }
    return newDates;
  }

  /**
   *  Translate the response into a readable object.
   * @returns {Object} returns a readable object
   * @private
   */
  static translateTrajStationsResp(resp) {
    const stations = [];
    for (let i = 0; i < resp.sts.length; i += 1) {
      const {
        sid: stationId,
        n: stationName,
        p: coordinates,
        at: arrivalTime,
        dt: departureTime,
        ap: arrivalDate,
        dp: departureDate,
        ad: arrivalDelay,
        dd: departureDelay,
        dot: noDropOff,
        put: noPickUp,
        c: cancelled,
        wa: wheelchairAccessible,
      } = resp.sts[i];

      stations.push({
        stationId,
        stationName,
        coordinates,
        arrivalTime: arrivalTime !== -1 ? arrivalDate * 1000 : null,
        departureTime: departureTime !== -1 ? departureDate * 1000 : null,
        arrivalDelay,
        departureDelay,
        noDropOff: !!noDropOff,
        noPickUp: !!noPickUp,
        cancelled: !!cancelled,
        wheelchairAccessible: !!wheelchairAccessible,
      });
    }

    const { n: operator, u: operatorUrl, tz: operatorTimeZone } = resp.a;

    const notOperatingDays = TrajservLayer.translateDates(resp.tt.n);
    const additionalOperatingDays = TrajservLayer.translateDates(resp.tt.p);
    const { t: operatingPeriod } = resp.tt;

    const {
      id,
      hs: destination,
      t: vehicleType,
      ln: longName,
      sn: shortName,
      wa: wheelchairAccessible,
      ba: bicyclesAllowed,
      rt: realTime,
      fid: feedsId,
      rid: routeIdentifier,
    } = resp;

    const backgroundColor = resp.c && `#${resp.c}`;
    const color = resp.tc && `#${resp.tc}`;

    return {
      id,
      destination,
      backgroundColor,
      color,
      vehicleType,
      routeIdentifier,
      longName,
      shortName,
      stations,
      wheelchairAccessible: !!wheelchairAccessible,
      bicyclesAllowed: !!bicyclesAllowed,
      realTime,
      feedsId,
      operatingInformations: {
        operatingPeriod,
        notOperatingDays,
        additionalOperatingDays,
      },
      operator,
      operatorUrl,
      operatorTimeZone,
    };
  }

  /**
   * Create a filter based on train and operator
   * @param {string} line
   * @param {string} route
   * @param {string} operator
   * @param {string} regexLine
   * @private
   */
  static createFilter(line, trip, operator, regexLine) {
    const filterList = [];

    if (!line && !trip && !operator && !regexLine) {
      return null;
    }

    if (regexLine) {
      const regexLineList =
        typeof regexLine === 'string' ? [regexLine] : regexLine;
      const lineFilter = (t) =>
        regexLineList.some((tr) => new RegExp(tr, 'i').test(t.name));
      filterList.push(lineFilter);
    }

    if (line) {
      const lineFiltersList = typeof line === 'string' ? line.split(',') : line;
      const lineList = lineFiltersList.map((l) =>
        l.replace(/\s+/g, '').toUpperCase(),
      );
      const lineFilter = (l) => {
        return lineList.some((filter) => filter === l.name.toUpperCase());
      };
      filterList.push(lineFilter);
    }

    if (trip) {
      const tripFilters = typeof trip === 'string' ? trip.split(',') : trip;
      const tripList = tripFilters.map((rt) => parseInt(rt, 10));
      const tripFilter = (t) => {
        const tripId = parseInt(t.routeIdentifier.split('.')[0], 10);
        return tripList.some((tr) => tr === tripId);
      };
      filterList.push(tripFilter);
    }

    if (operator) {
      const operatorList = typeof operator === 'string' ? [operator] : operator;
      const operatorFilter = (t) =>
        operatorList.some((op) => new RegExp(op, 'i').test(t.operator));
      filterList.push(operatorFilter);
    }

    return (t) => {
      for (let i = 0; i < filterList.length; i += 1) {
        if (!filterList[i](t)) {
          return false;
        }
      }
      return true;
    };
  }

  constructor(options = {}) {
    super({ ...options });

    this.options = options;
    this.url = options.url || 'https://api.geops.io/tracker/v1';
    this.showVehicleTraj =
      options.showVehicleTraj !== undefined ? options.showVehicleTraj : true;
    this.apiKey = options.apiKey;
    this.delayDisplay = options.delayDisplay || 300000;
    this.requestIntervalSeconds = 3;
    this.useDelayStyle = options.useDelayStyle || false;
    this.delayOutlineColor = options.delayOutlineColor || '#000000';
  }

  /**
   * Initialize the layer and listen to feature clicks.
   * @param {ol.map} map {@link https://openlayers.org/en/latest/apidoc/module-ol_Map-Map.html ol/Map)
   * @private
   */
  init(map) {
    super.init(map);

    if (!this.map) {
      return;
    }

    // Sort the trajectories.
    if (this.sortFc) {
      this.setSort(this.sortFc);
    } else if (this.useDelayStyle) {
      // Automatic sorting depending on delay, higher delay on top.
      this.setSort((a, b) => {
        if (a.delay === null) return 1;
        return a.delay < b.delay ? 1 : -1;
      });
    }
  }

  addTrackerFilters() {
    // Setting filters from the permalink.
    const parameters = qs.parse(window.location.search.toLowerCase());
    const lineParam = parameters[TrajservLayer.LINE_FILTER];
    const routeParam = parameters[TrajservLayer.ROUTE_FILTER];
    const opParam = parameters[TrajservLayer.OPERATOR_FILTER];
    const { regexPublishedLineName } = this.options;

    if (lineParam || routeParam || opParam || regexPublishedLineName) {
      this.filterFc = TrajservLayer.createFilter(
        lineParam ? lineParam.split(',') : undefined,
        routeParam ? routeParam.split(',') : undefined,
        opParam ? opParam.split(',') : undefined,
        regexPublishedLineName,
      );
    } else {
      this.filterFc = null;
    }

    this.setFilter(this.filterFc);
  }

  start() {
    if (!this.map) {
      return;
    }

    this.addTrackerFilters();

    super.start(this.map);
    this.startUpdateTrajectories();
    this.olEventsKeys = [
      ...this.olEventsKeys,
      this.map.on('singleclick', (e) => {
        if (!this.clickCallbacks.length) {
          return;
        }

        const [vehicle] = this.getVehiclesAtCoordinate(e.coordinate);
        const features = [];

        if (vehicle) {
          const geom = vehicle.coordinate
            ? new Point(vehicle.coordinate)
            : null;
          features.push(new Feature({ geometry: geom, ...vehicle }));

          if (features.length) {
            this.selectedVehicleId = features[0].get('id');
            this.journeyId = features[0].get('journeyIdentifier');
            this.fetchTrajectoryStations(this.selectedVehicleId).then((r) => {
              this.clickCallbacks.forEach((c) => c(r, this, e));
            });
          }
        } else {
          this.selectedVehicleId = null;
          this.olLayer.getSource().clear();
          this.clickCallbacks.forEach((c) => c(null, this, e));
        }
      }),
      this.map.on('moveend', () => {
        this.updateTrajectories();
        if (this.selectedVehicleId && this.journeyId) {
          this.highlightTrajectory();
        }
      }),
    ];
  }

  stop() {
    this.journeyId = null;
    this.stopUpdateTrajectories();
    this.abortFetchTrajectories();
    super.stop();
  }

  /**
   * Draw the trajectory as a line with points for each stop.
   * @param {Array} stationsCoords Array of station coordinates.
   * @param {<LineString|MultiLineString} lineGeometry A LineString or a MultiLineString.
   * @param {string} color The color of the line.
   * @private
   */
  drawFullTrajectory(stationsCoords, lineGeometry, color) {
    // Don't allow white lines, use red instead.
    const vehiculeColor = /#ffffff/i.test(color) ? '#ff0000' : color;
    const vectorSource = this.olLayer.getSource();
    vectorSource.clear();

    if (stationsCoords) {
      const geometry = new MultiPoint(stationsCoords);
      const aboveStationsFeature = new Feature(geometry);
      aboveStationsFeature.setStyle(
        new Style({
          zIndex: 1,
          image: new Circle({
            radius: 5,
            fill: new Fill({
              color: '#000000',
            }),
          }),
        }),
      );
      const belowStationsFeature = new Feature(geometry);
      belowStationsFeature.setStyle(
        new Style({
          zIndex: 4,
          image: new Circle({
            radius: 4,
            fill: new Fill({
              color: this.useDelayStyle ? '#a0a0a0' : vehiculeColor,
            }),
          }),
        }),
      );
      vectorSource.addFeatures([aboveStationsFeature, belowStationsFeature]);
    }

    const lineFeat = new Feature({
      geometry: lineGeometry,
    });
    lineFeat.setStyle([
      new Style({
        zIndex: 2,
        stroke: new Stroke({
          color: '#000000',
          width: 6,
        }),
      }),
      new Style({
        zIndex: 3,
        stroke: new Stroke({
          color: this.useDelayStyle ? '#a0a0a0' : vehiculeColor,
          width: 4,
        }),
      }),
    ]);
    vectorSource.addFeature(lineFeat);
  }

  /**
   * Fetch trajectories at given URL.
   * @param {string} url
   * @private
   */
  fetchTrajectories(url) {
    this.abortFetchTrajectories();
    this.abortController = new AbortController();
    const { signal } = this.abortController;
    return fetch(url, { signal })
      .then((data) => data.json())
      .catch((err) => {
        if (err.name === 'AbortError') {
          return;
        }
        // eslint-disable-next-line no-console
        console.warn('Fetch trajectories request failed: ', err);
      });
  }

  abortFetchTrajectories() {
    if (this.abortController) {
      this.abortController.abort();
    }
  }

  /**
   * Fetch stations information with a trajectory ID
   * @param {number} trajId The ID of the trajectory
   * @private
   */
  fetchTrajectoryStations(trajId) {
    const params = this.getUrlParams({
      id: trajId,
      time: getUTCTimeString(new Date()),
    });

    const url = `${this.url}/trajstations?${params}`;
    return fetch(url)
      .then((res) => {
        try {
          return res.json();
        } catch (err) {
          throw new Error(err);
        }
      })
      .then((resp) => {
        const trajStations = TrajservLayer.translateTrajStationsResp(resp);

        this.stationsCoords = [];
        trajStations.stations.forEach((station) => {
          this.stationsCoords.push(
            transformCoords(station.coordinates, 'EPSG:4326', 'EPSG:3857'),
          );
        });

        this.highlightTrajectory();
        return trajStations;
      });
  }

  highlightTrajectory() {
    this.fetchTrajectoryById(this.journeyId)
      .then((traj) => {
        const { p: multiLine, t, c } = traj;
        const lineCoords = [];
        multiLine.forEach((line) => {
          line.forEach((point) => {
            lineCoords.push([point.x, point.y]);
          });
        });

        this.drawFullTrajectory(
          this.stationsCoords,
          new LineString(lineCoords),
          c ? `#${c}` : getBgColor(t),
        );
      })
      .catch(() => {
        this.olLayer.getSource().clear();
      });
  }

  /**
   * Fetch trajectory information with a trajectory ID
   * @param {number} journeyId The gtfs ID of the trajectory.
   * @private
   */
  fetchTrajectoryById(journeyId) {
    const params = this.getUrlParams({
      id: journeyId,
      time: getUTCTimeString(new Date()),
    });

    const url = `${this.url}/trajectorybyid?${params}`;
    return fetch(url).then((res) => {
      try {
        return res.json();
      } catch (err) {
        throw new Error(err);
      }
    });
  }

  getApiKey() {
    return this.apiKey;
  }

  setApiKey(apiKey) {
    this.apiKey = apiKey;
  }

  /**
   * Returns the URL Parameters
   * @param {Object} extraParams
   * @returns {Object}
   * @private
   */
  getUrlParams(extraParams = {}) {
    const ext = this.map.getView().calculateExtent();
    const bbox = buffer(ext, getWidth(ext) / 10).join(',');
    const intervalMs = this.speed * 20000; // 20 seconds, arbitrary value, could be : (this.requestIntervalSeconds + 1) * 1000;
    const now = this.currTime;

    let diff = true;

    if (
      this.later &&
      now.getTime() > this.later.getTime() - 3000 * this.speed
    ) {
      diff = false;
    }
    if (
      !this.later ||
      !diff ||
      this.later.getTime() - now.getTime() > intervalMs
    ) {
      const later = new Date(now.getTime() + intervalMs);
      this.later = later;
    }

    const params = {
      ...extraParams,
      bbox,
      btime: getUTCTimeString(now),
      etime: getUTCTimeString(this.later),
      date: getDateString(now),
      rid: 1,
      a: 1,
      cd: 1,
      nm: 1,
      fl: 1,
      s: this.map.getView().getZoom() < 10 ? 1 : 0,
      z: this.map.getView().getZoom(),
      key: this.apiKey,
      // toff: this.currTime.getTime() / 1000,
    };

    // Allow to load only differences between the last request,
    // but currently the Tracker render method doesn't manage to render only diff.
    /* if (diff) {
      // Not working
      params.diff = this.lastRequestTime;
    } */

    return Object.keys(params)
      .map((k) => `${k}=${params[k]}`)
      .join('&');
  }

  /**
   * Start the update of trajectories.
   * @private
   */
  startUpdateTrajectories() {
    this.stopUpdateTrajectories();

    this.updateTrajectories();
    this.updateInterval = window.setInterval(() => {
      this.updateTrajectories();
    }, this.requestIntervalSeconds * 1000);
  }

  /**
   * Stop the update of trajectories.
   * @private
   */
  stopUpdateTrajectories() {
    clearInterval(this.updateInterval);
  }

  /**
   * Update the trajectories
   * @private
   */
  updateTrajectories() {
    this.fetchTrajectories(
      `${this.url}/trajectory_collection?${this.getUrlParams({
        attr_det: 1,
      })}`,
    ).then((data) => {
      if (!data) {
        return;
      }
      const trajectories = [];
      for (let i = 0; i < data.features.length; i += 1) {
        const traj = data.features[i];
        const geometry = new LineString(traj.geometry.coordinates);
        const {
          ID: id,
          ProductIdentifier: type,
          JourneyIdentifier: journeyIdentifier,
          PublishedLineName: name,
          RouteIdentifier: routeIdentifier,
          Operator: operator,
          TimeIntervals: timeIntervals,
          Color: color,
          TextColor: textColor,
          Delay: delay,
          Cancelled: cancelled,
        } = traj.properties;

        trajectories.push({
          id,
          type,
          name,
          color: color && `#${color}`,
          textColor: textColor && `#${textColor}`,
          delay,
          operator,
          journeyIdentifier,
          routeIdentifier,
          timeIntervals,
          geometry,
          cancelled,
        });
      }
      this.tracker.setTrajectories(trajectories);
    });
  }

  /**
   * Define the style of the vehicle.
   *
   * @param {Object} props Properties
   * @private
   */
  style(props) {
    const { type, name, id, color, textColor, delay, cancelled } = props;
    const z = Math.min(Math.floor(this.currentZoom || 1), 16);
    const hover = this.tracker.hoverVehicleId === id;
    const selected = this.selectedVehicleId === id;
    const key = `${z}${type}${name}${delay}${hover}${selected}`;

    if (!this.styleCache[key]) {
      let radius = getRadius(type, z);

      if (hover || selected) {
        radius += 5;
      }
      const margin = 1;
      const radiusDelay = radius + 2;
      const origin = radiusDelay + margin;

      const canvas = document.createElement('canvas');
      canvas.width = radiusDelay * 2 + margin * 2 + 100;
      canvas.height = radiusDelay * 2 + margin * 2;
      const ctx = canvas.getContext('2d');

      if (delay !== null) {
        // Draw delay background
        ctx.save();
        ctx.beginPath();
        ctx.arc(origin, origin, radiusDelay, 0, 2 * Math.PI, false);
        ctx.fillStyle = getDelayColor(delay, cancelled);
        ctx.filter = 'blur(1px)';
        ctx.fill();
        ctx.restore();
      }

      // Show delay if feature is hovered or if delay is above 5mins.
      if (hover || delay >= this.delayDisplay) {
        // Draw delay text
        ctx.save();
        ctx.textAlign = 'left';
        ctx.textBaseline = 'middle';
        ctx.font = `bold ${Math.max(
          14,
          Math.min(17, radius * 1.2),
        )}px arial, sans-serif`;
        ctx.fillStyle = getDelayColor(delay, cancelled);

        ctx.strokeStyle = this.delayOutlineColor;
        ctx.lineWidth = 1.5;
        ctx.strokeText(getDelayText(delay, cancelled), origin * 2, origin);
        ctx.fillText(getDelayText(delay, cancelled), origin * 2, origin);
        ctx.restore();
      }

      ctx.beginPath();
      ctx.arc(origin, origin, radius, 0, 2 * Math.PI, false);
      if (!this.useDelayStyle) {
        ctx.fillStyle = color || getBgColor(type);
        ctx.fill();
      } else {
        ctx.fillStyle = getDelayColor(delay, cancelled);
        ctx.fill();
      }

      ctx.lineWidth = 1;
      ctx.strokeStyle = '#000000';
      ctx.stroke();

      const markerSize = radius * 2;
      if (radius > 10) {
        const shortname =
          type === 'Rail' && name.length > 3 ? name.substring(0, 2) : name;
        const fontSize = Math.max(radius, 10);
        const textSize = getTextSize(ctx, markerSize, shortname, fontSize);

        ctx.textBaseline = 'middle';
        ctx.textAlign = 'center';
        ctx.fillStyle = !this.useDelayStyle
          ? textColor || getTextColor(type)
          : '#000000';
        ctx.font = `bold ${textSize}px Arial`;
        ctx.fillText(shortname, origin, origin);
      }
      this.styleCache[key] = canvas;
    }

    return this.styleCache[key];
  }
}

TrajservLayer.LINE_FILTER = 'publishedlinename';
TrajservLayer.ROUTE_FILTER = 'tripnumber';
TrajservLayer.OPERATOR_FILTER = 'operator';

export default TrajservLayer;
