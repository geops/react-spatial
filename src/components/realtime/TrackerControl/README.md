#

This demonstrates the use of TrackerControl.

```jsx
import React from 'react';
import BasicMap from 'react-spatial/components/maps/BasicMap';
import Layer from 'react-spatial/layers/Layer';
import TileLayer from 'ol/layer/Tile';
import OSMSource from 'ol/source/OSM';
import TrajservLayer from 'react-spatial/layers/TrajservLayer';
import TrackerControl from 'react-spatial/components/realtime/TrackerControl';

// The `apiKey` used here is for demonstration purposes only.
// Please get your own api key at https://developer.geops.io/.
const trackerLayer = new TrajservLayer({
  apiKey: window.apiKey,
});

const layers = [
  new Layer({
    name: 'Layer',
    olLayer: new TileLayer({
      source: new OSMSource(),
    }),
  }),
  trackerLayer,
];

function TrackerControlExample() {
  return (
    <>
      <BasicMap center={[951560, 6002550]} zoom={14} layers={layers} />
      <TrackerControl trackerLayer={trackerLayer} />
    </>
  );
}

<TrackerControlExample />;
```
