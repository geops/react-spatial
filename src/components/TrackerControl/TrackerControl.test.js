import React from 'react';
import renderer from 'react-test-renderer';
import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import TrajservLayer from '../../layers/TrajservLayer';
import TrackerControl from '.';

configure({ adapter: new Adapter() });

describe('TrackerControl', () => {
  test('matches snapshots.', () => {
    const trackerLayer = new TrajservLayer();
    const component = renderer.create(
      <TrackerControl trackerLayer={trackerLayer} />,
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
