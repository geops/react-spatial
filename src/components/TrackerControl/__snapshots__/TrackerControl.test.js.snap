// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TrackerControl matches snapshots. 1`] = `
<div
  className="rt-tracker-control"
>
  <div
    aria-label="reset date"
    className="rt-control-button"
    onClick={[Function]}
    onKeyPress={[Function]}
    role="button"
    tabIndex={0}
    title="reset date"
  >
    <svg
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 512 512"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M256 56c110.532 0 200 89.451 200 200 0 110.532-89.451 200-200 200-110.532 0-200-89.451-200-200 0-110.532 89.451-200 200-200m0-48C119.033 8 8 119.033 8 256s111.033 248 248 248 248-111.033 248-248S392.967 8 256 8zm0 168c-44.183 0-80 35.817-80 80s35.817 80 80 80 80-35.817 80-80-35.817-80-80-80z"
      />
    </svg>
  </div>
  <div
    aria-label="speed down"
    className="rt-control-button"
    onClick={[Function]}
    onKeyPress={[Function]}
    role="button"
    tabIndex={0}
    title="speed down"
  >
    <svg
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 512 512"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M11.5 280.6l192 160c20.6 17.2 52.5 2.8 52.5-24.6V96c0-27.4-31.9-41.8-52.5-24.6l-192 160c-15.3 12.8-15.3 36.4 0 49.2zm256 0l192 160c20.6 17.2 52.5 2.8 52.5-24.6V96c0-27.4-31.9-41.8-52.5-24.6l-192 160c-15.3 12.8-15.3 36.4 0 49.2z"
      />
    </svg>
  </div>
  <div
    aria-label="speed reset"
    className="rt-control-button"
    onClick={[Function]}
    onKeyPress={[Function]}
    role="button"
    tabIndex={0}
    title="speed reset"
  >
    <svg
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 448 512"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M424.4 214.7L72.4 6.6C43.8-10.3 0 6.1 0 47.9V464c0 37.5 40.7 60.1 72.4 41.3l352-208c31.4-18.5 31.5-64.1 0-82.6z"
      />
    </svg>
  </div>
  <div
    aria-label="speed up"
    className="rt-control-button"
    onClick={[Function]}
    onKeyPress={[Function]}
    role="button"
    tabIndex={0}
    title="speed up"
  >
    <svg
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 512 512"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M500.5 231.4l-192-160C287.9 54.3 256 68.6 256 96v320c0 27.4 31.9 41.8 52.5 24.6l192-160c15.3-12.8 15.3-36.4 0-49.2zm-256 0l-192-160C31.9 54.3 0 68.6 0 96v320c0 27.4 31.9 41.8 52.5 24.6l192-160c15.3-12.8 15.3-36.4 0-49.2z"
      />
    </svg>
  </div>
  <div
    className="rt-tracker-speed"
  >
    <svg
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 512 512"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M256 64C132.3 64 32 164.2 32 287.9c0 57 21.3 109 56.3 148.5 1.1 1.2 2.1 2.4 3.2 3.5C96.3 445 103 448 110 448c7.1 0 13.8-3 18.6-8.2 31.6-34.4 77-55.9 127.4-55.9s95.8 21.6 127.4 55.9c4.8 5.2 11.5 8.2 18.6 8.2 7 0 13.7-2.9 18.5-8.1 1.1-1.2 2.1-2.3 3.2-3.5 35-39.5 56.3-91.5 56.3-148.5C480 164.2 379.7 64 256 64zm-14 45.9c0-7.7 6.3-14 14-14s14 6.3 14 14v36c0 7.7-6.3 14-14 14s-14-6.3-14-14v-36zm-128 192H78c-7.7 0-14-6.3-14-14s6.3-14 14-14h36c7.7 0 14 6.3 14 14s-6.3 14-14 14zm51.5-104.5c-5.4 5.4-14.4 5.4-19.8 0L120.2 172c-5.4-5.4-5.4-14.4 0-19.8s14.4-5.4 19.8 0l25.5 25.5c5.4 5.4 5.4 14.3 0 19.7zm160.6 34.5l-47.5 75.5c-1.9 2.6-4.3 5.1-7 7-13.5 9.7-32.3 6.5-42-7s-6.5-32.3 7-42l75.5-47.5c3.4-2.4 8.1-2.5 11.7 0 4.5 3.2 5.5 9.5 2.3 14zm40.2-34.5c-5.4 5.4-14.4 5.4-19.8 0s-5.4-14.4 0-19.8l25.5-25.5c5.4-5.4 14.4-5.4 19.8 0s5.4 14.4 0 19.8l-25.5 25.5zM434 301.9h-36c-7.7 0-14-6.3-14-14s6.3-14 14-14h36c7.7 0 14 6.3 14 14s-6.3 14-14 14z"
      />
    </svg>
    1
  </div>
</div>
`;
